File: nimo-plm-web/src/api/ecn.ts
import apiClient from './client';
import { ApiResponse, Product, User } from '@/types';

export interface ECN {
  id: string;
  code: string;
  title: string;
  product_id: string;
  change_type: string;
  urgency: string;
  status: string;
  reason: string;
  description: string;
  impact_analysis: string;
  technical_plan: string;
  planned_date: string;
  completion_rate: number;
  approval_mode: string;
  sop_impact: Record<string, any>;
  requested_by: string;
  requested_at: string;
  approved_by: string;
  approved_at: string;
  rejection_reason: string;
  implemented_by: string;
  implemented_at: string;
  feishu_approval_code: string;
  feishu_instance_code: string;
  created_at: string;
  updated_at: string;
  product?: Product;
  requester?: User;
  approver?: User;
  implementer?: User;
  affected_items?: ECNAffectedItem[];
  approvals?: ECNApproval[];
  tasks?: ECNTask[];
}

export interface ECNAffectedItem {
  id: string;
  ecn_id: string;
  item_type: string;
  item_id: string;
  material_code: string;
  material_name: string;
  affected_bom_ids: Record<string, any>;
  before_value: Record<string, any>;
  after_value: Record<string, any>;
  change_description: string;
  created_at: string;
}

export interface ECNApproval {
  id: string;
  ecn_id: string;
  approver_id: string;
  sequence: number;
  status: string;
  decision: string;
  comment: string;
  decided_at: string;
  created_at: string;
  approver?: User;
}

export interface ECNTask {
  id: string;
  ecn_id: string;
  type: string;
  title: string;
  description: string;
  assignee_id: string;
  due_date: string;
  status: string;
  completed_at: string;
  completed_by: string;
  metadata: Record<string, any>;
  sort_order: number;
  created_at: string;
  updated_at: string;
  assignee?: User;
}

export interface ECNHistory {
  id: string;
  ecn_id: string;
  action: string;
  user_id: string;
  detail: Record<string, any>;
  created_at: string;
  user?: User;
}

export interface ECNStats {
  pending_approval: number;
  executing: number;
  month_created: number;
  month_closed: number;
}

export interface ECNListResponse {
  items: ECN[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface CreateECNRequest {
  title: string;
  product_id: string;
  change_type: string;
  urgency?: string;
  reason: string;
  description?: string;
  impact_analysis?: string;
  technical_plan?: string;
  planned_date?: string;
  approval_mode?: string;
  sop_impact?: Record<string, any>;
  affected_items?: Partial<ECNAffectedItem>[];
  approver_ids?: string[];
}

export const ecnApi = {
  list: async (params?: { status?: string; product_id?: string; keyword?: string; change_type?: string; urgency?: string; page?: number; page_size?: number }): Promise<ECNListResponse> => {
    const searchParams = new URLSearchParams();
    if (params?.status) searchParams.set('status', params.status);
    if (params?.product_id) searchParams.set('product_id', params.product_id);
    if (params?.keyword) searchParams.set('keyword', params.keyword);
    if (params?.change_type) searchParams.set('change_type', params.change_type);
    if (params?.urgency) searchParams.set('urgency', params.urgency);
    if (params?.page) searchParams.set('page', params.page.toString());
    if (params?.page_size) searchParams.set('page_size', params.page_size.toString());
    const query = searchParams.toString();
    const response = await apiClient.get<ApiResponse<ECNListResponse>>(`/ecns${query ? `?${query}` : ''}`);
    return response.data.data;
  },

  get: async (id: string): Promise<ECN> => {
    const response = await apiClient.get<ApiResponse<ECN>>(`/ecns/${id}`);
    return response.data.data;
  },

  create: async (data: CreateECNRequest): Promise<ECN> => {
    const response = await apiClient.post<ApiResponse<ECN>>('/ecns', data);
    return response.data.data;
  },

  update: async (id: string, data: Partial<ECN>): Promise<ECN> => {
    const response = await apiClient.put<ApiResponse<ECN>>(`/ecns/${id}`, data);
    return response.data.data;
  },

  submit: async (id: string): Promise<ECN> => {
    const response = await apiClient.post<ApiResponse<ECN>>(`/ecns/${id}/submit`, {});
    return response.data.data;
  },

  approve: async (id: string, data: { comment?: string }): Promise<ECN> => {
    const response = await apiClient.post<ApiResponse<ECN>>(`/ecns/${id}/approve`, data);
    return response.data.data;
  },

  reject: async (id: string, data: { reason: string }): Promise<ECN> => {
    const response = await apiClient.post<ApiResponse<ECN>>(`/ecns/${id}/reject`, data);
    return response.data.data;
  },

  implement: async (id: string): Promise<ECN> => {
    const response = await apiClient.post<ApiResponse<ECN>>(`/ecns/${id}/implement`, {});
    return response.data.data;
  },

  // 统计
  getStats: async (): Promise<ECNStats> => {
    const response = await apiClient.get<ApiResponse<ECNStats>>('/ecns/stats');
    return response.data.data;
  },

  getMyPending: async (): Promise<{ items: ECN[] }> => {
    const response = await apiClient.get<ApiResponse<{ items: ECN[] }>>('/ecns/my-pending');
    return response.data.data;
  },

  // 受影响项
  listAffectedItems: async (id: string): Promise<ECNAffectedItem[]> => {
    const response = await apiClient.get<ApiResponse<ECNAffectedItem[]>>(`/ecns/${id}/affected-items`);
    return response.data.data;
  },

  addAffectedItem: async (id: string, data: Partial<ECNAffectedItem>): Promise<ECNAffectedItem> => {
    const response = await apiClient.post<ApiResponse<ECNAffectedItem>>(`/ecns/${id}/affected-items`, data);
    return response.data.data;
  },

  updateAffectedItem: async (id: string, itemId: string, data: Partial<ECNAffectedItem>): Promise<ECNAffectedItem> => {
    const response = await apiClient.put<ApiResponse<ECNAffectedItem>>(`/ecns/${id}/affected-items/${itemId}`, data);
    return response.data.data;
  },

  removeAffectedItem: async (id: string, itemId: string): Promise<void> => {
    await apiClient.delete(`/ecns/${id}/affected-items/${itemId}`);
  },

  // 审批
  listApprovals: async (id: string): Promise<ECNApproval[]> => {
    const response = await apiClient.get<ApiResponse<ECNApproval[]>>(`/ecns/${id}/approvals`);
    return response.data.data;
  },

  addApprover: async (id: string, data: { approver_id: string; sequence?: number }): Promise<ECNApproval> => {
    const response = await apiClient.post<ApiResponse<ECNApproval>>(`/ecns/${id}/approvers`, data);
    return response.data.data;
  },

  // 执行任务
  listTasks: async (id: string): Promise<{ items: ECNTask[] }> => {
    const response = await apiClient.get<ApiResponse<{ items: ECNTask[] }>>(`/ecns/${id}/tasks`);
    return response.data.data;
  },

  createTask: async (id: string, data: Partial<ECNTask>): Promise<ECNTask> => {
    const response = await apiClient.post<ApiResponse<ECNTask>>(`/ecns/${id}/tasks`, data);
    return response.data.data;
  },

  updateTask: async (id: string, taskId: string, data: Partial<ECNTask>): Promise<ECNTask> => {
    const response = await apiClient.put<ApiResponse<ECNTask>>(`/ecns/${id}/tasks/${taskId}`, data);
    return response.data.data;
  },

  // BOM变更应用
  applyBOMChanges: async (id: string): Promise<void> => {
    await apiClient.post(`/ecns/${id}/apply-bom-changes`, {});
  },

  // 操作历史
  listHistory: async (id: string): Promise<{ items: ECNHistory[] }> => {
    const response = await apiClient.get<ApiResponse<{ items: ECNHistory[] }>>(`/ecns/${id}/history`);
    return response.data.data;
  },
};

File: internal/plm/entity/ecn.go
package entity

import (
	"time"
)

// ECN 工程变更通知
type ECN struct {
	ID                   string     `json:"id" gorm:"primaryKey;size:32"`
	Code                 string     `json:"code" gorm:"size:64;not null;uniqueIndex"`
	Title                string     `json:"title" gorm:"size:256;not null"`
	ProductID            string     `json:"product_id" gorm:"size:32;not null"`
	ChangeType           string     `json:"change_type" gorm:"size:32;not null"`
	Urgency              string     `json:"urgency" gorm:"size:16;not null;default:medium"`
	Status               string     `json:"status" gorm:"size:16;not null;default:draft"`
	Reason               string     `json:"reason" gorm:"type:text;not null"`
	Description          string     `json:"description" gorm:"type:text"`
	ImpactAnalysis       string     `json:"impact_analysis" gorm:"type:text"`
	TechnicalPlan        string     `json:"technical_plan" gorm:"type:text"`
	PlannedDate          *time.Time `json:"planned_date"`
	CompletionRate       int        `json:"completion_rate" gorm:"default:0"`
	ApprovalMode         string     `json:"approval_mode" gorm:"size:16;default:serial"`
	SOPImpact            JSONB      `json:"sop_impact" gorm:"type:jsonb"`
	RequestedBy          string     `json:"requested_by" gorm:"size:32;not null"`
	RequestedAt          *time.Time `json:"requested_at"`
	ApprovedBy           *string    `json:"approved_by" gorm:"size:32"`
	ApprovedAt           *time.Time `json:"approved_at"`
	RejectionReason      string     `json:"rejection_reason" gorm:"type:text"`
	ImplementedBy        *string    `json:"implemented_by" gorm:"size:32"`
	ImplementedAt        *time.Time `json:"implemented_at"`
	FeishuApprovalCode   string     `json:"feishu_approval_code" gorm:"size:64"`
	FeishuInstanceCode   string     `json:"feishu_instance_code" gorm:"size:64"`
	CreatedAt            time.Time  `json:"created_at"`
	UpdatedAt            time.Time  `json:"updated_at"`

	// 关联
	Product       *Product           `json:"product,omitempty" gorm:"foreignKey:ProductID"`
	Requester     *User              `json:"requester,omitempty" gorm:"foreignKey:RequestedBy"`
	Approver      *User              `json:"approver,omitempty" gorm:"foreignKey:ApprovedBy"`
	Implementer   *User              `json:"implementer,omitempty" gorm:"foreignKey:ImplementedBy"`
	AffectedItems []ECNAffectedItem  `json:"affected_items,omitempty" gorm:"foreignKey:ECNID"`
	Approvals     []ECNApproval      `json:"approvals,omitempty" gorm:"foreignKey:ECNID"`
	Tasks         []ECNTask          `json:"tasks,omitempty" gorm:"foreignKey:ECNID"`
}

func (ECN) TableName() string {
	return "ecns"
}

// ECNAffectedItem ECN受影响项目
type ECNAffectedItem struct {
	ID                string    `json:"id" gorm:"primaryKey;size:32"`
	ECNID             string    `json:"ecn_id" gorm:"size:32;not null"`
	ItemType          string    `json:"item_type" gorm:"size:32;not null"`
	ItemID            string    `json:"item_id" gorm:"size:32;not null"`
	MaterialCode      string    `json:"material_code" gorm:"size:64"`
	MaterialName      string    `json:"material_name" gorm:"size:256"`
	AffectedBOMIDs    JSONB     `json:"affected_bom_ids" gorm:"type:jsonb"`
	BeforeValue       JSONB     `json:"before_value" gorm:"type:jsonb"`
	AfterValue        JSONB     `json:"after_value" gorm:"type:jsonb"`
	ChangeDescription string    `json:"change_description" gorm:"type:text"`
	CreatedAt         time.Time `json:"created_at"`

	// 关联
	ECN *ECN `json:"ecn,omitempty" gorm:"foreignKey:ECNID"`
}

func (ECNAffectedItem) TableName() string {
	return "ecn_affected_items"
}

// ECNApproval ECN审批记录
type ECNApproval struct {
	ID         string     `json:"id" gorm:"primaryKey;size:32"`
	ECNID      string     `json:"ecn_id" gorm:"size:32;not null"`
	ApproverID string     `json:"approver_id" gorm:"size:32;not null"`
	Sequence   int        `json:"sequence" gorm:"not null"`
	Status     string     `json:"status" gorm:"size:16;not null;default:pending"`
	Decision   string     `json:"decision" gorm:"size:16"`
	Comment    string     `json:"comment" gorm:"type:text"`
	DecidedAt  *time.Time `json:"decided_at"`
	CreatedAt  time.Time  `json:"created_at"`

	// 关联
	ECN      *ECN  `json:"ecn,omitempty" gorm:"foreignKey:ECNID"`
	Approver *User `json:"approver,omitempty" gorm:"foreignKey:ApproverID"`
}

func (ECNApproval) TableName() string {
	return "ecn_approvals"
}

// ECNTask ECN执行任务
type ECNTask struct {
	ID          string     `json:"id" gorm:"primaryKey;size:32"`
	ECNID       string     `json:"ecn_id" gorm:"size:32;not null;index"`
	Type        string     `json:"type" gorm:"size:32;not null"`
	Title       string     `json:"title" gorm:"size:256;not null"`
	Description string     `json:"description" gorm:"type:text"`
	AssigneeID  string     `json:"assignee_id" gorm:"size:32"`
	DueDate     *time.Time `json:"due_date"`
	Status      string     `json:"status" gorm:"size:16;not null;default:pending"`
	CompletedAt *time.Time `json:"completed_at"`
	CompletedBy *string    `json:"completed_by" gorm:"size:32"`
	Metadata    JSONB      `json:"metadata" gorm:"type:jsonb"`
	SortOrder   int        `json:"sort_order" gorm:"not null;default:0"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`

	// 关联
	ECN      *ECN  `json:"ecn,omitempty" gorm:"foreignKey:ECNID"`
	Assignee *User `json:"assignee,omitempty" gorm:"foreignKey:AssigneeID"`
}

func (ECNTask) TableName() string {
	return "ecn_tasks"
}

// ECNHistory ECN操作历史
type ECNHistory struct {
	ID        string    `json:"id" gorm:"primaryKey;size:32"`
	ECNID     string    `json:"ecn_id" gorm:"size:32;not null;index"`
	Action    string    `json:"action" gorm:"size:32;not null"`
	UserID    string    `json:"user_id" gorm:"size:32;not null"`
	Detail    JSONB     `json:"detail" gorm:"type:jsonb"`
	CreatedAt time.Time `json:"created_at"`

	// 关联
	ECN  *ECN  `json:"ecn,omitempty" gorm:"foreignKey:ECNID"`
	User *User `json:"user,omitempty" gorm:"foreignKey:UserID"`
}

func (ECNHistory) TableName() string {
	return "ecn_histories"
}

// ECN状态常量
const (
	ECNStatusDraft       = "draft"
	ECNStatusPending     = "pending"
	ECNStatusApproved    = "approved"
	ECNStatusRejected    = "rejected"
	ECNStatusExecuting   = "executing"
	ECNStatusClosed      = "closed"
	ECNStatusImplemented = "implemented"
	ECNStatusCancelled   = "cancelled"
)

// ECN变更类型
const (
	ECNChangeTypeDesign   = "design"
	ECNChangeTypeMaterial = "material"
	ECNChangeTypeProcess  = "process"
	ECNChangeTypeSpec     = "spec"
	ECNChangeTypeDocument = "document"
)

// ECN紧急程度
const (
	ECNUrgencyLow      = "low"
	ECNUrgencyMedium   = "medium"
	ECNUrgencyHigh     = "high"
	ECNUrgencyCritical = "critical"
)

// ECN受影响项目类型
const (
	ECNAffectedTypeBOMItem   = "bom_item"
	ECNAffectedTypeMaterial  = "material"
	ECNAffectedTypeDocument  = "document"
	ECNAffectedTypeDrawing   = "drawing"
)

// ECN审批决策
const (
	ECNApprovalDecisionApprove = "approve"
	ECNApprovalDecisionReject  = "reject"
)

// ECN审批模式
const (
	ECNApprovalModeSerial   = "serial"
	ECNApprovalModeParallel = "parallel"
)

// ECN执行任务类型
const (
	ECNTaskTypeBOMUpdate       = "bom_update"
	ECNTaskTypeDrawingUpdate   = "drawing_update"
	ECNTaskTypeSupplierNotify  = "supplier_notify"
	ECNTaskTypeInventoryHandle = "inventory_handle"
	ECNTaskTypeDocUpdate       = "doc_update"
	ECNTaskTypeSOPUpdate       = "sop_update"
)

// ECN执行任务状态
const (
	ECNTaskStatusPending    = "pending"
	ECNTaskStatusInProgress = "in_progress"
	ECNTaskStatusCompleted  = "completed"
	ECNTaskStatusSkipped    = "skipped"
)

// ECN历史动作
const (
	ECNHistoryCreated       = "created"
	ECNHistoryUpdated       = "updated"
	ECNHistorySubmitted     = "submitted"
	ECNHistoryApproved      = "approved"
	ECNHistoryRejected      = "rejected"
	ECNHistoryExecuting     = "executing"
	ECNHistoryTaskCompleted = "task_completed"
	ECNHistoryClosed        = "closed"
	ECNHistoryBOMApplied    = "bom_applied"
)
